Properly integrate tool XML schema with docs

nsoranzo opened the pull request:
- Generate intermediate `schema.rst` file instead of html
- Move generation to `doc/Makefile`
- Use `pandoc` to generate table of contents and fragment ids.
- Remove nested inline markup (not supported by reStructuredText)
- Fix https and other links.

Also:
- Fix a bunch of errors reported by Sphinx (many more remaining)
- Add and use local images for build runner doc
- Fix flake8 errors and Python3 compatibility for `doc/patch.py` xref. #1715 
- Remove 2 unused files

Ping @jmchilton @martenson 

![screenshot of Galaxy tool XML doc](https://cloud.githubusercontent.com/assets/4924623/19660597/63b36114-9a28-11e6-801b-87d885975e9e.png)

nsoranzo committed:
Fix flake8 errors and Python3 compatibility for doc/patch.py

nsoranzo committed:
Move png in the right directory

Broken in #2543.

nsoranzo committed:
Remove unused files

nsoranzo committed:
Add and use local images for build runner doc

nsoranzo committed:
Properly integrate tool XML schema with docs

Generate intermediate rst file instead of html
Move generation to doc/Makefile
Use pandoc to generate table of contents and fragment ids.
Remove nested inline markup (not supported by reStructuredText)
Fix https and other links.

nsoranzo committed:
Fix a bunch of errors reported by Sphinx

martenson commented on doc/Makefile:
```diff
@@ -43,44 +44,49 @@ help:
 	@echo "  updaterst  to update sphinx rst to reflect code structure changes"
 
+schema.md: parse_gx_xsd.py schema_template.md ../lib/galaxy/tools/xsd/galaxy.xsd ## Build Github-flavored Markdown from Galaxy Tool XSD (expects libxml in environment)
+	python parse_gx_xsd.py schema_template.md ../lib/galaxy/tools/xsd/galaxy.xsd > $@
```
@nsoranzo why is this called a second time (with `python`)? What does `$@` refer to - is it the make target (`schema.md`)?

nsoranzo replied:
It is called only one time, the first line should be read as: `schema.md` should be regenerated if any of the prerequisites (`parse_gx_xsd.py`, `schema_template.md`, `../lib/galaxy/tools/xsd/galaxy.xsd`) has changed after this file.

`$@` is a special `make` variable which is substituted with the make target, as you were correctly supposing.

martenson commented:
note: This will require changes to docs-building tasks at https://jenkins.galaxyproject.org/  when merged.

nsoranzo commented:
qunit build failure is unrelated, it's broken since the merge of #3040.

nsoranzo commented:
Force-pushed to fix merge conflict with #3087.

jmchilton commented:
Looks great for the most part - thanks a ton @nsoranzo!

martenson commented:
This really does require changes to doc building tasks in Jenkins...

nsoranzo commented:
@martenson Yep, hopefully it should be a simplification!

jmchilton commented:
For the record - current task definition is:

```
./scripts/common_startup.sh --no-create-venv
pip install sphinx
pip install sphinx_rtd_theme
pip install lxml
pip install recommonmark
make docs-schema-html

sed -e '/  <\/style>/ i\
body{\
  width: 95%;\
  margin-left: auto !important;\
  margin-right: auto !important;\
}\
table{\
  width: 100% !important;\
}\
table, th, td{\
  padding: 0.33em !important;\
  border-style: dotted !important;\
  border-width: 1px !important;\
  border-color: grey !important;\
  border-collapse: collapse !important;\
}' -e '/  <\/style>/ a\
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">' doc/schema.html > doc/schema.html.modified

sed -i -e 's|<br />| |g' doc/schema.html.modified
sed -i -e 's|gt;</code></a>|gt;</code></a><br />|g' doc/schema.html.modified
sed -i -e 's|<title>|<title>Galaxy tool schema documentation|' doc/schema.html.modified
cd doc
make html
mv schema.html.modified build/html/dev/schema.html
```

martenson commented:
clean & smooth in the Galaxy way!

nsoranzo commented:
I think that you just need to substitute:

```
make docs-schema-html

sed -e '/  <\/style>/ i\
body{\
  width: 95%;\
  margin-left: auto !important;\
  margin-right: auto !important;\
}\
table{\
  width: 100% !important;\
}\
table, th, td{\
  padding: 0.33em !important;\
  border-style: dotted !important;\
  border-width: 1px !important;\
  border-color: grey !important;\
  border-collapse: collapse !important;\
}' -e '/  <\/style>/ a\
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">' doc/schema.html > doc/schema.html.modified

sed -i -e 's|<br />| |g' doc/schema.html.modified
sed -i -e 's|gt;</code></a>|gt;</code></a><br />|g' doc/schema.html.modified
sed -i -e 's|<title>|<title>Galaxy tool schema documentation|' doc/schema.html.modified
cd doc
make html
mv schema.html.modified build/html/dev/schema.html
```

with:

```
make -C doc/ html
```

jmchilton commented:
https://docs.galaxyproject.org/en/latest/dev/schema.html

I went with `make docs` instead. Any clue why the TOC doesn't show up at the top of the document on docs.galaxyproject.org but it appears locally?

I guess it is on the side - that might be enough?

nsoranzo commented:
Old pandoc maybe?

nsoranzo commented:
Answering to myself, it's not pandoc, table of contents is present in the .rst : https://docs.galaxyproject.org/en/latest/_sources/dev/schema.txt

Must be a problem with sphinx then.

martenson commented:
I think we do not want the TOC in the left menu either, that looks really messy and not useful.

martenson commented:
The state the schema docs is in now is a regression to how it looked before. The TOC at the top of the page was one of the best features I think. We are going to use this on Monday in SLC admin training. :/

martenson commented on doc/schema_template.md:
```diff
@@ -10,11 +10,9 @@ If you find a bug please report it [here](https://github.com/galaxyproject/galax
+[Planemo](https://planemo.readthedocs.io/en/latest/writing.html) features a
 number of tutorials on building Galaxy tools that would better serve that purpose.
 
-$toc
```
could this be the reason the toc is missing from top of the document? @nsoranzo

nsoranzo replied:
No, this doesn't work now that we go through rst. As mentioned above, the rst created by pandoc contains the TOC command at the top:

``` rst
.. contents::
  :depth: 2.0
..
```

but Sphinx for unknown reasons (old version maybe) ignores it on Jenkins while it works locally for both @jmchilton and me.

Please note that the commands above are general rst, while Sphix has also its own syntax: http://www.sphinx-doc.org/en/1.4.8/markup/toctree.html

nsoranzo replied:
Found the reason why Sphinx is ignoring it: if `depth` is a float, Sphinx silently skips the toc creation. The above RST command should have been:

``` rst
.. contents::
  :depth: 2
..
```

According to https://pandoc.org/changelog.txt , this has been fixed in Pandoc 1.14:

```
* RST writer:

    + Fixed toc depth in RST writer.  Previously the depth was being
      rendered as a floating point number with a decimal point.
```

I'll prepare a workaround.

nsoranzo replied:
Should be fixed by https://github.com/galaxyproject/galaxy/pull/3135
